src/test/java/org/apache/ibatis/submitted/sqlprovider/OurSqlBuilder.java

```diff
@@ -213,4 +218,107 @@ public String buildSelectByIdAndNameMultipleParamAndProviderContext(final Intege
+    Class<?> declaringClass = mapperMethod.getDeclaringClass();
+    Class<?> mapperClass = providerContext.getMapperType();
+
+    Type[] types = mapperClass.getGenericInterfaces();
+    for (Type type : types) {
+      if (type instanceof ParameterizedType) {
+        ParameterizedType t = (ParameterizedType) type;
+        if (t.getRawType() == declaringClass || mapperClass.isAssignableFrom((Class<?>) t.getRawType())) {
+          Class<?> returnType = (Class<?>) t.getActualTypeArguments()[0];
+          return returnType;
+        }
+      }
+    }
+    throw new RuntimeException("The interface ["
+        + mapperClass.getCanonicalName() + "] must specify a generic type.");
+  }
+
+  private Map<String, String> getColumnMap(ProviderContext context){
+    Class entityClass = getEntityClass(context);
+    Field[] fields = entityClass.getDeclaredFields();
+    Map<String, String> columnMap = new LinkedHashMap<String, String>();
+    for (Field field : fields) {
+      BaseMapper.Column column = field.getAnnotation(BaseMapper.Column.class);
+      if(column != null){
+        String columnName = column.value();
+        if(columnName == null || columnName.length() == 0){
+          columnName = field.getName();
+        }
+        columnMap.put(columnName, field.getName());
+      }
+    }
+    if(columnMap.size() == 0){
+      throw new RuntimeException("There is no field in the class ["
+          + entityClass.getCanonicalName() + "] that specifies the @BaseMapper.Column annotation.");
+    }
+    return columnMap;
+  }
+
+  public String buildInsertSelective(ProviderContext context) {
+    final String tableName = context.getMapperType().getAnnotation(BaseMapper.Meta.class).tableName();
+    Map<String, String> columnMap = getColumnMap(context);
+    StringBuffer sqlBuffer = new StringBuffer();
+    sqlBuffer.append("<script>");
+    sqlBuffer.append("insert into ");
+    sqlBuffer.append(tableName);
+    sqlBuffer.append("<trim prefix=\"(\" suffix=\")\" suffixOverrides=\",\">");
+    for (Map.Entry<String, String> entry : columnMap.entrySet()) {
+      sqlBuffer.append("<if test=\"").append(entry.getValue()).append(" != null\">");
+      sqlBuffer.append(entry.getKey()).append(",");
+      sqlBuffer.append("</if>");
+    }
+    sqlBuffer.append("</trim>");
+    sqlBuffer.append("<trim prefix=\"VALUES (\" suffix=\")\" suffixOverrides=\",\">");
+    for (String field : columnMap.values()) {
+      sqlBuffer.append("<if test=\"").append(field).append(" != null\">");
+      sqlBuffer.append("#{").append(field).append("} ,");
+      sqlBuffer.append("</if>");
+    }
+    sqlBuffer.append("</trim>");
+    sqlBuffer.append("</script>");
+    return sqlBuffer.toString();
+  }
+
+  public String buildUpdateSelective(ProviderContext context) {
+    final String tableName = context.getMapperType().getAnnotation(BaseMapper.Meta.class).tableName();
+    Map<String, String> columnMap = getColumnMap(context);
```

Comment:
a `MUCH` better approach here would be to use a template event if it's custom and put the template file in the resources directory